### ISSUE TITLE
Support for dependency locking, do not use absolute paths

### ISSUE THREAD

**Issue by AlexanderBartash** [JetBrains/intellij-platform-gradle-plugin#1778]
### What happened?

Here https://github.com/JetBrains/intellij-platform-gradle-plugin/blob/33af9bdd1d0d8b996b911781ab23013dd14997f4/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformDependenciesHelper.kt#L1097

The plugin does
```kotlin
private val Path.hash
    get() = (absolutePathString().hashCode().absoluteValue % 1000).toString()
```
which then is used to create dependencies in Gradle, also in transforms.

```
Property
    Path.hash
Usages in All Places  (7 usages found)
    Production  (7 usages found)
        IntelliJPlatformGradlePlugin.main  (7 usages found)
            IntelliJPlatformDependenciesHelper.kt  (7 usages found)
                IntelliJPlatformDependenciesHelper  (7 usages found)
                    collectDependencies  (2 usages found)
                        833 val version = "${plugin.pluginVersion}+${artifactPath.hash}"
                        861 val version = "$buildNumber+${platformPath.hash}"
                    createIntelliJPlatformBundledModule  (1 usage found)
                        795 val version = baseVersion.orElse(productInfo.map { it.version }).map { "$it+${platformPath.hash}" }.get()
                    createIntelliJPlatformBundledPlugin  (1 usage found)
                        763 val version = baseVersion.orElse(productInfo.map { it.version }).map { "$it+${artifactPath.hash}" }.get()
                    createIntelliJPlatformLocal  (1 usage found)
                        700 val version = "${localProductInfo.buildNumber}+${artifactPath.hash}"
                    createIntelliJPlatformLocalPlugin  (1 usage found)
                        897 val version = (plugin.pluginVersion ?: "0.0.0") + "+" + artifactPath.hash
                    createJetBrainsRuntimeLocal  (1 usage found)
                        939 val version = javaVendorVersion.removePrefix(name).trim('-') + "+" + artifactPath.hash

```

To understand the below it is important to understand how Gradle generates hashes for the transform cache. E.g. 
`/home/user/.gradle/caches/8.10.2/transforms/3eefd31b055e78c772b75249fc0f0276/transformed/ideaIU-242-EAP-SNAPSHOT/plugins/JavaEE` that transform hash `3eefd31b055e78c772b75249fc0f0276` is MD5 generated by https://github.com/gradle/gradle/blob/6236f2642795dc014e11eb5065c5e0d3e1e376bc/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformWorkspaceIdentity.java#L26 which may take into account many things but never absolute path:
- artifact name
- file hash (content)
- directory tree content hash
- Not really sure, but seems correct: artifact dependency coordinates or the way it was added to the dependencies, meaning that if we add a usual dependency on a lib it is one hash, but if we then add another with strict dependency version it will be another hash.

Docs https://docs.gradle.org/current/userguide/artifact_transforms.html#artifact_transforms_with_caching

This is bad for a number of reasons:
- I may be wrong, but it may be not following the convention https://maven.apache.org/guides/getting-started/index.html#what-is-a-snapshot-version It seems like you can add something before `o x.(y+1)-SNAPSHOT`  but not at the end.
- That absolute path usually points somewhere ilike `/home/user/.gradle/caches/8.10.2/transforms/3eefd31b055e78c772b75249fc0f0276/transformed/ideaIU-242-EAP-SNAPSHOT/plugins/JavaEE` which we hash and make a part of artifact coordinates. It can hurt things like https://docs.gradle.org/current/userguide/build_cache_debugging.html#caching_relocation_test I am not really sure if it breaks the build cache, but it does messes up transforms cache for sure, see below.
- Hashes will be different on different machines, which might make it difficult to commit some things into VCS is those dependencies get listed.
- With certain conditions met it breaks Gradle dependency locking https://docs.gradle.org/current/userguide/dependency_locking.html
- This one may be related maybe not, but https://docs.gradle.org/current/userguide/dependency_verification.html is not usable at all. It may be a separate issue, I will provide details as well. See https://github.com/JetBrains/intellij-platform-gradle-plugin/issues/1779

### Steps to reproduce
See tests in https://github.com/JetBrains/intellij-platform-gradle-plugin/pull/1780
https://github.com/AlexanderBartash/intellij-platform-plugin-template/tree/dep-lock-bug

To reproduce the problem run:
`./gradlew --write-locks clean build`

this will generate a bunch of lock files, with strict rules on our all dependencies, including transitive.

`./gradlew build`

It will complain that a bunch (looks somewhat random, but probably is not) dependencies changed, because
now they have a different +number hash ending, which changed.

I think it happens, probably, because Gradle dependency locking works by simply adding to the project all those strict dependencies from the lock files to the project and lets dependency conflict resolution do the work
https://docs.gradle.org/current/userguide/dependency_constraints_conflicts.html#resolving_version_conflicts

If I am correct, then, it means, that we will probably have two platform dependencies, which may cause ExtractorTransformer to run again, since the version is different, creating new plugin dependencies, with new hashes, which makes dependency conflict resolution impossible. Or it may be that the presence of the dependency constraint causes the transform to run again, creating new artifact version (since the hash of the absolute path), which causes dependency locks to fail.

By default Gradle expects different behavior from transforms:
https://docs.gradle.org/current/userguide/artifact_transforms.html#executing_artifact_transforms
> can be run in parallel
> are cacheable
> are reusable (if separate resolutions used by different tasks require the same transform to be executed on the same artifacts, the transform results will be cached and shared)

I do not quite understand why some dependencies break that way and some don't.

```
FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache state could not be cached: field `provider` of `org.gradle.internal.serialize.codecs.core.ProviderBackedFileCollectionSpec` bean found in field `element` of `java.util.Collections$SingletonList` bean found in field `elements` of `org.gradle.internal.serialize.codecs.core.ResolutionBackedFileCollectionSpec` bean found in field `__classpathSnapshot__` of `org.jetbrains.kotlin.gradle.tasks.KotlinCompile$ClasspathSnapshotProperties` bean found in field `__classpathSnapshotProperties__` of task `:compileKotlin` of type `org.jetbrains.kotlin.gradle.tasks.KotlinCompile`: error writing value of type 'org.gradle.api.internal.provider.DefaultProvider'
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve bundledPlugin:com.intellij.database:2023.3.8+994.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:com.intellij.database' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:2023.3.8+994'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:{strictly 2023.3.8+947}' because of the following reason: dependency was locked to version '2023.3.8+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.javaee:2023.3.8+703' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.persistence:2023.3.8+420' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.hibernate:233.15619.7+77' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.javaee.jpa:233.15619.7+982' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:Tomcat:2023.3.8+267' (default) --> 'bundledPlugin:com.intellij.javaee.app.servers.integration:233.15619.7+879' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'

   > Could not resolve bundledPlugin:com.intellij.diagram:2023.3.8+670.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:com.intellij.diagram' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.diagram:2023.3.8+670'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.diagram:{strictly 2023.3.8+247}' because of the following reason: dependency was locked to version '2023.3.8+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:2023.3.8+994' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+670'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.persistence:2023.3.8+420' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:2023.3.8+994' (default) --> 'bundledPlugin:JavaScript:233.15619.7+835' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.javaee.jpa:233.15619.7+982' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.microservices.jvm:233.15619.7+433' (default) --> 'bundledPlugin:com.jetbrains.restClient:233.15619.7+532' (default) --> 'bundledPlugin:com.intellij.microservices.ui:233.15619.7+324' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'

   > Could not resolve bundledPlugin:Tomcat:2023.3.8+267.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:Tomcat' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:Tomcat:2023.3.8+267'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:Tomcat:{strictly 2023.3.8+84}' because of the following reason: dependency was locked to version '2023.3.8+84'

   > Could not resolve bundledPlugin:org.jetbrains.idea.maven:2023.3.8+493.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:org.jetbrains.idea.maven' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:org.jetbrains.idea.maven:2023.3.8+493'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:org.jetbrains.idea.maven:{strictly 2023.3.8+222}' because of the following reason: dependency was locked to version '2023.3.8+222'

> There are 7 more failures with identical causes.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 2s
Configuration cache entry discarded due to serialization error.
```

### Gradle IntelliJ Plugin version

2.1.0

### Gradle version

8.10.2

### Operating System

Linux

**Comment by AlexanderBartash** [comment]
I was able to confirm that ExtractorTransformer runs twice when dependency locks are enabled, generated & actually used.

What is described below was done without my fix: https://github.com/JetBrains/intellij-platform-gradle-plugin/pull/1780

So I did the next experiment in https://github.com/AlexanderBartash/intellij-platform-plugin-template/tree/dep-lock-bug
1. I killed Gradle daemons and removed `.gradle` in home.
2. Then: `./gradlew build`
[build-ExtractorTransformer-0.log](https://github.com/user-attachments/files/17214223/build-ExtractorTransformer-0.log)
ExtractorTransformer run for the first time and created
`/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517/transformed/ideaIU-2023.3.8`
```
Selected primary task 'build' from project :
Resource missing. [HTTP GET: https://repo.maven.apache.org/maven2/com/jetbrains/intellij/idea/ideaIU/2023.3.8/ideaIU-2023.3.8.pom]
Downloading https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/releases/com/jetbrains/intellij/idea/ideaIU/2023.3.8/ideaIU-2023.3.8.pom to /home/sasha/.gradle/.tmp/gradle_download4070131720423878460bin
Downloading https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/releases/com/jetbrains/intellij/idea/ideaIU/2023.3.8/ideaIU-2023.3.8.zip to /home/sasha/.gradle/.tmp/gradle_download15370965056771427860bin
Transforming ideaIU-2023.3.8.zip (com.jetbrains.intellij.idea:ideaIU:2023.3.8) with ExtractorTransformer
Caching disabled for ExtractorTransformer: /home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip because:
  Caching not enabled.
[org.jetbrains.intellij.platform] Extracting archive '/home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip' to directory '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Resolving the content directory in '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] The content directory is '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Extracting to '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8' completed.
```
3. Then I generated locks with: `./gradlew --write-locks build`
[build-ExtractorTransformer-1.log](https://github.com/user-attachments/files/17214187/build-ExtractorTransformer-1.log)
Transform cache was reused, as expected.
4. Then I tried to build again: `./gradlew build`
I expect this build to succeed because we just generated locks and they should be valid. But ExtractorTransformer run again and created another `/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e/transformed/ideaIU-2023.3.8`. Which due to absolute path hashing changes dependencies version string and fails the build, because versions for the same artifacts from the previous step have different hashes, because there the cache was reused.
[build-ExtractorTransformer-2.log](https://github.com/user-attachments/files/17214188/build-ExtractorTransformer-2.log)
```log
Task name matched 'build'
Selected primary task 'build' from project :
Loaded state for lock with ID 'jetbrainsRuntime'
Loaded state for lock with ID 'intellijPlatformDependency'
Loaded state for lock with ID 'intellijPlatformDependency'
Transforming ideaIU-2023.3.8.zip (com.jetbrains.intellij.idea:ideaIU:2023.3.8) with ExtractorTransformer
Caching disabled for ExtractorTransformer: /home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip because:
  Caching not enabled.
[org.jetbrains.intellij.platform] Extracting archive '/home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip' to directory '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Resolving the content directory in '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] The content directory is '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Extracting to '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8' completed.
```
```
FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache state could not be cached: field `provider` of `org.gradle.internal.serialize.codecs.core.ProviderBackedFileCollectionSpec` bean found in field `element` of `java.util.Collections$SingletonList` bean found in field `elements` of `org.gradle.internal.serialize.codecs.core.ResolutionBackedFileCollectionSpec` bean found in field `__classpathSnapshot__` of `org.jetbrains.kotlin.gradle.tasks.KotlinCompile$ClasspathSnapshotProperties` bean found in field `__classpathSnapshotProperties__` of task `:compileKotlin` of type `org.jetbrains.kotlin.gradle.tasks.KotlinCompile`: error writing value of type 'org.gradle.api.internal.provider.DefaultProvider'
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve bundledPlugin:com.intellij.database:2023.3.8+994.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:com.intellij.database' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:2023.3.8+994'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:{strictly 2023.3.8+947}' because of the following reason: dependency was locked to version '2023.3.8+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.javaee:2023.3.8+703' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.persistence:2023.3.8+420' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.hibernate:233.15619.7+77' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.javaee.jpa:233.15619.7+982' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:Tomcat:2023.3.8+267' (default) --> 'bundledPlugin:com.intellij.javaee.app.servers.integration:233.15619.7+879' (default) --> 'bundledPlugin:com.intellij.database:233.15619.7+947'

   > Could not resolve bundledPlugin:com.intellij.diagram:2023.3.8+670.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:com.intellij.diagram' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.diagram:2023.3.8+670'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.diagram:{strictly 2023.3.8+247}' because of the following reason: dependency was locked to version '2023.3.8+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:2023.3.8+994' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+670'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.persistence:2023.3.8+420' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.database:2023.3.8+994' (default) --> 'bundledPlugin:JavaScript:233.15619.7+835' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.javaee.jpa:233.15619.7+982' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:com.intellij.spring:2023.3.8+706' (default) --> 'bundledPlugin:com.intellij.microservices.jvm:233.15619.7+433' (default) --> 'bundledPlugin:com.jetbrains.restClient:233.15619.7+532' (default) --> 'bundledPlugin:com.intellij.microservices.ui:233.15619.7+324' (default) --> 'bundledPlugin:com.intellij.diagram:233.15619.7+247'

   > Could not resolve bundledPlugin:Tomcat:2023.3.8+267.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:Tomcat' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:Tomcat:2023.3.8+267'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:Tomcat:{strictly 2023.3.8+84}' because of the following reason: dependency was locked to version '2023.3.8+84'

   > Could not resolve bundledPlugin:org.jetbrains.idea.maven:2023.3.8+493.
     Required by:
         root project :
      > Cannot find a version of 'bundledPlugin:org.jetbrains.idea.maven' that satisfies the version constraints:
           Dependency path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:org.jetbrains.idea.maven:2023.3.8+493'
           Constraint path 'org.jetbrains.plugins.template:IntelliJ Platform Plugin Template:2.0.1' --> 'bundledPlugin:org.jetbrains.idea.maven:{strictly 2023.3.8+222}' because of the following reason: dependency was locked to version '2023.3.8+222'

> There are 7 more failures with identical causes.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 32s
Configuration cache entry discarded due to serialization error.
```

I tried it with the fix https://github.com/JetBrains/intellij-platform-gradle-plugin/pull/1780 as well and ExtractorTransformer runs twice as well, the only difference is that now there are no hashes in the versions and the last build was successful.
Logs:
[pr1780-build-ExtractorTransformer-0.log](https://github.com/user-attachments/files/17214402/pr1780-build-ExtractorTransformer-0.log)
[pr1780-build-ExtractorTransformer-1.log](https://github.com/user-attachments/files/17214403/pr1780-build-ExtractorTransformer-1.log)
[pr1780-build-ExtractorTransformer-2.log](https://github.com/user-attachments/files/17214404/pr1780-build-ExtractorTransformer-2.log)

**Comment by AlexanderBartash** [comment]
> I do not quite understand why some dependencies break that way and some don't.

I checked a few dependencies, and it seems that it can be explained again by that generated 3 number hash from artifact absolute path. Gradle seems to pick an artifact with the biggest number. But if it happens that the dependency constraint get a lower number it cases build failure.

All magic has been explained now, except why sometimes the transformations are cached and when Gradle adds constraints they can not be reused. There is a theory, explained below. Quote from the end of this comment:

> I have no idea why Gradle cached it before, despite me trying everything to stop it, but does not reuse the cache now. It may be that we simply should remove `@DisableCachingByDefault(because = "Not worth caching")` and implement caching and it will reuse it. It may be so, because, that cache might be something of "last resort cache" kind, meaning that Gradle sees that nothing changed, so why run the transform again. But then we add a bunch of constrains and obviously the cache can not be reused, especially since the transformations has `@DisableCachingByDefault(because = "Not worth caching")`.

Relevant documentation:
-  https://docs.gradle.org/current/userguide/rich_versions.html
-  https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html

> strict dependencies don’t suffer an ordering problem: they are applied transitively to the subgraph, and it doesn’t matter in which order dependencies are declared.

> Using a strict version must be carefully considered, in particular by library authors. As the producer, a strict version will effectively behave like a force: the version declaration takes precedence over whatever is found in the transitive dependency graph. In particular, a strict version will override any other strict version on the same module found transitively.

And an example:

This does not fail the build because here the constraint does an upgrade.
```
dependencies {
    implementation("commons-io:commons-io:2.16.1")
    implementation("commons-io:commons-io") {
        version {
            strictly("2.17.0")
        }
    }
}
```
This fails the build, because here the constraint does an downgrade. It would work if "2.17.0" would have come from a transitive dependency, instead of being duplicated as an another declared dependency in the same project.
```
dependencies {
    implementation("commons-io:commons-io:2.17.0")
    implementation("commons-io:commons-io") {
        version {
            strictly("2.16.1")
        }
    }
}
```

So what happens is:
- Assume we have removed all caches.
- Assume we have a dependency on `bundledPlugin:Tomcat` and `com.jetbrains.intellij.idea:ideaIU:2023.3.8`.
- Keep in mind that we do not specify versions of bundledPlugin we depend on, it gets assigned by this plugin each time we run `./gradlew build`. Probably, it could get cached in the configuration cache if that is enabled, so that the same logic for adding the same dependency does not have to run when nothing relevant changed.

1. We run `./gradlew build` ExtractorTransformer runs for the first time and creates 
```
/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8
```
```
Selected primary task 'build' from project :
Resource missing. [HTTP GET: https://repo.maven.apache.org/maven2/com/jetbrains/intellij/idea/ideaIU/2023.3.8/ideaIU-2023.3.8.pom]
Downloading https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/releases/com/jetbrains/intellij/idea/ideaIU/2023.3.8/ideaIU-2023.3.8.pom to /home/sasha/.gradle/.tmp/gradle_download4070131720423878460bin
Downloading https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/releases/com/jetbrains/intellij/idea/ideaIU/2023.3.8/ideaIU-2023.3.8.zip to /home/sasha/.gradle/.tmp/gradle_download15370965056771427860bin
Transforming ideaIU-2023.3.8.zip (com.jetbrains.intellij.idea:ideaIU:2023.3.8) with ExtractorTransformer
Caching disabled for ExtractorTransformer: /home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip because:
  Caching not enabled.
[org.jetbrains.intellij.platform] Extracting archive '/home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip' to directory '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Resolving the content directory in '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] The content directory is '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Extracting to '/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-9020d4ce-416c-4c55-8c12-e0e7d791771b/transformed/ideaIU-2023.3.8' completed.
```
 2. The build is still running. This plugin generates a dependency for our declated `bundledPlugin:Tomcat` and it happens that `84` is a hash for:
```
/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517/transformed/ideaIU-2023.3.8/plugins/javaee-appServers-tomcat
```

So Gradle gets a dependency  `bundledPlugin:Tomcat:2023.3.8+84`

3. We run a new build `./gradlew build --write-locks`, the previous steps are repeated (or configuration cache is used), but this time the transform cache is reused, the locks are created, including for `bundledPlugin:Tomcat:2023.3.8+84`.

This time we have 84 again because the cache was reused, we get the same absolute path to the artifact, thus the same hash.

Despite that transforms from this plugin are annotated by `@DisableCachingByDefault` the cache is still reused, even if I do `./gradlew build --no-build-cache --no-configuration-cache --no-watch-fs` and kill the daemon the log says:
```
Transforming ideaIU-2023.3.8.zip (com.jetbrains.intellij.idea:ideaIU:2023.3.8) with ExtractorTransformer
```
But I see that they do nothing because other logs from ExtractorTransformer are missing and it is quite heavy and takes time to run. 

Even with debug logging it is clear that it does not run because it run for 1ms (these logs are from another build, the hashes might not match the example, it is just to prove the point):

```
2024-10-02T00:10:55.085+0300 [DEBUG] [org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.DependencyGraphBuilder] Visiting configuration com.jetbrains.intellij.idea:ideaIU:2023.3.8(runtime).
2024-10-02T00:10:55.085+0300 [DEBUG] [org.gradle.internal.locking.DefaultDependencyLockingProvider] Loaded state for lock with ID 'intellijPlatformDependency', state is: [com.jetbrains.intellij.idea:ideaIU:2023.3.8]
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.api.internal.artifacts.ivyservice.resolveengine.oldresult.TransientConfigurationResultsBuilder] Flushing resolved configuration data in Binary store in /home/sasha/.gradle/.tmp/gradle14720832304734858937.bin. Wrote root 1555.
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Completing Build operation 'Resolve dependencies of :intellijPlatformDependency'
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Build operation 'Resolve dependencies of :intellijPlatformDependency' completed
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.component.model.LoggingAttributeMatchingExplanationBuilder] Candidate com.jetbrains.intellij.idea:ideaIU:2023.3.8 configuration runtime doesn't match attributes {intellijPlatformExtracted=true}
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Build operation 'Resolve files of configuration ':intellijPlatformDependency'' started
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Build operation 'Execute transform' started
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.api.internal.artifacts.ivyservice.ivyresolve.CachingModuleComponentRepository] Found artifact 'ideaIU-2023.3.8.zip (com.jetbrains.intellij.idea:ideaIU:2023.3.8)' in resolver cache: /home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip
2024-10-02T00:10:55.086+0300 [INFO] [org.gradle.api.internal.artifacts.transform.TransformStep] Transforming ideaIU-2023.3.8.zip (com.jetbrains.intellij.idea:ideaIU:2023.3.8) with ExtractorTransformer
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Build operation 'Identifying work' started
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Completing Build operation 'Identifying work'
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Build operation 'Identifying work' completed
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Completing Build operation 'Execute transform'
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Build operation 'Execute transform' completed
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Completing Build operation 'Resolve files of configuration ':intellijPlatformDependency''
2024-10-02T00:10:55.086+0300 [DEBUG] [org.gradle.internal.operations.DefaultBuildOperationRunner] Build operation 'Resolve files of configuration ':intellijPlatformDependency'' completed
```

4. We run `./gradlew build` again, but this time we have two Tomcat dependencies:
 - `bundledPlugin:Tomcat` -- our declated dependency, no version since we do not specify version for them.
 - `bundledPlugin:Tomcat:{strict 2023.3.8+84}` -- added by Gradle from the dependency lock.
 
Despite that transforms according to the Gradle [docs](https://docs.gradle.org/current/userguide/artifact_transforms.html#executing_artifact_transforms):
> can be run in parallel
> are cacheable
> are reusable (if separate resolutions used by different tasks require the same transform to be executed on the same artifacts, the transform results will be cached and shared)

Gradle decides to run ExtractorTransformer again and parallelism is not the issue here, because I tried `--no-parallel` with no luck.
 ```
 Task name matched 'build'
Selected primary task 'build' from project :
Loaded state for lock with ID 'jetbrainsRuntime'
Loaded state for lock with ID 'intellijPlatformDependency'
Loaded state for lock with ID 'intellijPlatformDependency'
Transforming ideaIU-2023.3.8.zip (com.jetbrains.intellij.idea:ideaIU:2023.3.8) with ExtractorTransformer
Caching disabled for ExtractorTransformer: /home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip because:
  Caching not enabled.
[org.jetbrains.intellij.platform] Extracting archive '/home/sasha/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIU/2023.3.8/7cb6ca200b78b921b30a70b5057b952b5bfebd2a/ideaIU-2023.3.8.zip' to directory '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Resolving the content directory in '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] The content directory is '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8'.
[org.jetbrains.intellij.platform] Extracting to '/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e-9e355738-303e-4811-82d3-c11e9a4485b7/transformed/ideaIU-2023.3.8' completed.
 ```

I have no idea why Gradle cached it before, despite me trying everything to stop it, but does not reuse the cache now. It may be that we simply should remove `@DisableCachingByDefault(because = "Not worth caching")` and implement caching and it will reuse it. It may be so, because, that cache might be something of "last resort cache" kind, meaning that Gradle sees that nothing changed, so why run the transform again. But then we add a bunch of constrains and obviously the cache can not be reused, especially since the transformations has `@DisableCachingByDefault(because = "Not worth caching")`.

5. Since the transform run again, we get different artifact path:
```
/home/sasha/.gradle/caches/8.10.2/transforms/47266ca8ccefa5fcf2e6b5b95ae5e09e/transformed/ideaIU-2023.3.8/plugins/javaee-appServers-tomcat
 ```
So we get a different hash, which happens to be 267. And the plugin adds a dependency `bundledPlugin:Tomcat:2023.3.8+267`.

6. Now we have that situation when we have two declared dependencies:
- `bundledPlugin:Tomcat:2023.3.8+267` -- added by the plugin, but this time with a different hash.
- `bundledPlugin:Tomcat:{strict 2023.3.8+84}` -- from the dependency lock, which did not change because there was no reason for it.

And the build fails, because, as I've explained at the beginning of this comment, in two declared dependencies on a project, a strict constraint can upgrade but can not downgrade the version.

All this happens with all dependencies, it is just sometimes happens an upgrade and Gradle lets it be, but fails downgrades.

**Comment by AlexanderBartash** [comment]
Yeah, it actually makes sense I think. If caching is implemented for ExtractorTransformer & probably CollectorTransformer they won't run again when the constraints are added, because probably Gradle can not figure it out for us.

**Comment by AlexanderBartash** [comment]
Also disabled caching for ExtractorTransformer is bad because e.g. ideaIU-2023.3.8 2.8 GiB and ideaIU-242-EAP-SNAPSHOT 3.4 GiB. Copying gigabytes of files on a whim does not sound like a good idea. Especially when the input a single zip file which probably can be easily hashed, and outputs are not really expected to change I think.

**Comment by AlexanderBartash** [comment]
Okay, so the thing which causes the ExtractorTransformer to run again are dependencies added to the buildscript classpath.

I.e. this creates `buildscript-gradle.lockfile` which then is loaded by Gradle and constraints are added to the `classpath` configuration. And when we run build the next time Gradle does not reuse the cache because of it.
```
buildscript {
    dependencyLocking {
        lockAllConfigurations()
    }
}
```

I can even reproduce it without dependencies locking by:
1. Clear caches, kill demons.
2. `./gradlew build` creates 
`/home/sasha/.gradle/caches/8.10.2/transforms/2f4561d46bc638866dfa4bcb9a564517-4cd301fa-d135-4a6f-8899-25d3af640bf9/transformed/ideaIU-2023.3.8`
```
buildscript {
    dependencies {
//        classpath("commons-codec:commons-codec:1.17.0")
//        classpath("commons-codec:commons-codec") {
//            version {
//                strictly("1.17.0")
//            }
//        }
//        classpath("commons-codec:commons-codec") {
//            version {
//                strictly("1.17.1")
//            }
//        }
    }
}
```
3. `./gradlew build` creates 
`/home/sasha/.gradle/caches/8.10.2/transforms/7c5bcfc2a6f1aa609d29dc6e519bd5f9/transformed/ideaIU-2023.3.8`
```
buildscript {
    dependencies {
        classpath("commons-codec:commons-codec:1.17.0")
//        classpath("commons-codec:commons-codec") {
//            version {
//                strictly("1.17.0")
//            }
//        }
//        classpath("commons-codec:commons-codec") {
//            version {
//                strictly("1.17.1")
//            }
//        }
    }
}
```

4. `./gradlew build` does not create anything. So apparently constraints which do not change anything do not cause it.

```
buildscript {
    dependencies {
        classpath("commons-codec:commons-codec:1.17.0")
        classpath("commons-codec:commons-codec") {
            version {
                strictly("1.17.0")
            }
        }
//        classpath("commons-codec:commons-codec") {
//            version {
//                strictly("1.17.1")
//            }
//        }
    }
}
```
5. `./gradlew build` creates
`/home/sasha/.gradle/caches/8.10.2/transforms/848c4940f966d2b84a0180ee3d959424-92bf986f-129c-44c2-bc15-5dbe797c21a1/transformed/ideaIU-2023.3.8`
```
buildscript {
    dependencies {
        classpath("commons-codec:commons-codec:1.17.0")
//        classpath("commons-codec:commons-codec") {
//            version {
//                strictly("1.17.0")
//            }
//        }
        classpath("commons-codec:commons-codec") {
            version {
                strictly("1.17.1")
            }
        }
    }
}
```

**Comment by AlexanderBartash** [comment]
https://github.com/JetBrains/intellij-platform-gradle-plugin/issues/1601

**Comment by AlexanderBartash** [comment]
See https://github.com/JetBrains/intellij-platform-gradle-plugin/issues/1779#issuecomment-2395500297
